src/tracer.js

```diff
@@ -0,0 +1,218 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import assert from 'assert';
+import * as constants from './constants.js';
+import * as opentracing from 'opentracing';
+import myLocalIp from 'my-local-ip';
+import NoopReporter from './reporters/noop_reporter.js';
+import NoopSampler from './samplers/noop_sampler.js';
+import pjson from '../package.json';
+import Span from './span.js';
+import SpanContext from './span_context.js';
+import TextMapCodec from './propagators/text_map_codec.js';
+import * as thrift from './thrift.js';
+import Utils from './util.js';
+
+export default class Tracer {
+    _serviceName: string;
+    _reporter: Reporter;
+    _sampler: Sampler;
+    _host: Endpoint;
+    _tracerTags: any;
+    _injectors: any;
+    _extractors: any;
+
+    constructor(serviceName: string,
+            reporter: Reporter,
+            sampler: Sampler,
+            port: number  = 0,
+            tracerTags: any = {}) {
+        tracerTags.jaegerClient = `Node-${pjson.version}`;
+        this._tracerTags = tracerTags
+        this._serviceName = serviceName;
+        this._reporter = reporter || new NoopReporter();
+        this._sampler = sampler || new NoopSampler();
+        this._host = Utils.createEndpoint(serviceName, myLocalIp(), port);
+        this._injectors = {};
+        this._extractors = {};
+
+        let textCodec = new TextMapCodec(false);
+        this.registerInjector(opentracing.FORMAT_TEXT_MAP, textCodec);
+        this.registerExtractor(opentracing.FORMAT_TEXT_MAP, textCodec);
+
+        let httpCodec = new TextMapCodec(true);
+        this.registerInjector(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+        this.registerExtractor(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+    }
+
+    _createSpan(operationName: string,
+                spanContext: SpanContext,
+                isClient: boolean,
+                peer: Endpoint): Span {
+
+        let span = new Span(
+            this,
+            operationName,
+            spanContext,
+            Utils.getTimestampMicros(),
+            isClient,
+            peer
+        );
+
+        span.log({
+            'event': isClient ? thrift.CLIENT_SEND : thrift.SERVER_RECV,
+            'timestamp': span._timestamp
+        });
```

Comment:
I think we should move these to reporter. They are artifacts of Zipkin representation, we don't need them to be in the main tracer code.